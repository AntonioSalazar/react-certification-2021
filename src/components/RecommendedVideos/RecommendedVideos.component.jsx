import React from 'react';
import { Link, useLocation } from 'react-router-dom';
import './RecommendedVideos.styles.css';
import { TitleSpan } from './styledComponents';

const RecommendedVideos = ({ videos }) => {
  const recommendedVideos = videos.slice(0, 15);
  const { pathname } = useLocation();
  const currentPath = pathname.split('/');
  return (
    <>
      <div className="recommended-video__container">
        {recommendedVideos.map((video) => (
          <Link
            to={
              currentPath[1] === 'video'
                ? `/video/${video.id.videoId}`
                : `/favorite-video/${video.id}`
            }
            style={{ textDecoration: 'none', color: '#000' }}
            key={currentPath[1] === 'video' ? video.id.videoId : video.id}
          >
            <div className="recommended-video__card">
              <div className="recommended-video__card-img">
                <img
                  src={video.snippet.thumbnails.medium.url}
                  style={{ width: '100%' }}
                  alt={video.snippet.title}
                />
              </div>
              <div className="recommended-video__card-text">
                <TitleSpan>{video.snippet.title}</TitleSpan>
              </div>
            </div>
            <hr style={{ width: '80%', margin: '0 auto', color: '#060B26' }} />
          </Link>
        ))}
      </div>
    </>
  );
};

export default RecommendedVideos;
