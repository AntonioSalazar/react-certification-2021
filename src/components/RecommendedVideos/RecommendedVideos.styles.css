.recommended-video__card {
  width: 100%;
  display: flex;
  margin-bottom: 1rem;
  margin-top: 1rem;
}

.recommended-video__card-img {
  width: 40%;
  margin-left: 2rem;
}

.recommended-video__card-text {
  width: 60%;
  display: flex;
  align-items: center;
  font-size: 2rem;
}
