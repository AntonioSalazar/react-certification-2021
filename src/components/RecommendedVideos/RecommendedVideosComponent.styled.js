import styled from "styled-components";

export const TitleSpan = styled.span`
    font-size: 1.4rem;
    justify-content: left;
    width: 100%;
    display: inline-block;
`