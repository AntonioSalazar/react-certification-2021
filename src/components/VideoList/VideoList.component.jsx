import React from 'react';
import styled from 'styled-components';

const VideoCardDiv = styled.div`
  display: flex;
  flex-direction: column;
  width: 30rem;
  height: 35rem;
  border-radius: 0.4rem;
  margin: 2rem;
  border: 1px #dad4d4 solid;
  transition: all 0.3s ease;
  &:hover {
    box-shadow: 10px 10px 24px 0px rgba(0, 0, 0, 0.75);
  }
`;

const VideoImageDiv = styled.div`
  height: 140px;
`;

const TitleSpan = styled.span`
  font-size: 1.6rem;
  font-weight: bold;
  margin: 2rem;
`;

const DescriptionSpan = styled.span`
  font-size: 1.4rem;
  margin-bottom: 2rem;
  text-align: left;
  margin: 2rem;
`;

const VideoList = ({ video }) => {
  const { title, description, thumbnails } = video;

  return (
    <VideoCardDiv>
      <VideoImageDiv
        style={{
          backgroundImage: `url(${thumbnails.medium.url})`,
          backgroundSize: 'cover',
          backgroundRepeat: 'no-repeat',
          backgroundPosition: 'center',
          borderTopRightRadius: '0.4rem',
          borderTopLeftRadius: '0.4rem',
        }}
      ></VideoImageDiv>
      <TitleSpan>{title}</TitleSpan>
      <DescriptionSpan>{description ? description : 'watch video...'}</DescriptionSpan>
    </VideoCardDiv>
  );
};

export default VideoList;
