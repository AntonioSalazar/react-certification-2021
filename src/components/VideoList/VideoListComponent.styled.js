import styled from 'styled-components';

export const VideoCardDivLight = styled.div`
  display: flex;
  flex-direction: column;
  width: 30rem;
  height: 35rem;
  border-radius: 0.4rem;
  margin: 2rem;
  border: 1px var(--light-gray) solid;
  transition: all 0.3s ease;
  &:hover {
      box-shadow: 10px 10px 24px 0px var(--black);
    }
    
    @media only screen and (max-width: 670px) {
        width: 80%;
    }
`;
VideoCardDivLight.displayName = 'VideoCardDivLight';

export const VideoCardDivDark= styled.div`
  display: flex;
  flex-direction: column;
  width: 30rem;
  height: 35rem;
  border-radius: 0.4rem;
  margin: 2rem;
  border: 1px var(--light-gray) solid;
  transition: all 0.3s ease;
  &:hover {
      -webkit-box-shadow: 5px 5px 15px 5px var(--black); 
      box-shadow: 5px 5px 15px 5px var(--light-gray) ;
    }
    
    @media only screen and (max-width: 670px) {
        width: 80%;
    }
`;
VideoCardDivDark.displayName = 'VideoCardDivDark';

export const VideoImageDiv = styled.div`
  height: 140px;
`;
VideoImageDiv.displayName = 'VideoImageDiv';

export const TitleSpan = styled.span`
  font-size: 1.6rem;
  font-weight: bold;
  margin: 2rem;
`;
TitleSpan.displayName = 'TitleSpan';

export const DescriptionSpan = styled.span`
  font-size: 1.4rem;
  margin-bottom: 2rem;
  text-align: left;
  margin: 2rem;
  `;
DescriptionSpan.displayName = 'DescriptionSpan';

