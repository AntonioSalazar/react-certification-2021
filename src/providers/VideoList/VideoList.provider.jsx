import React, { createContext, useEffect, useState } from 'react';
export const VideoListContext = createContext();
const VideoListProvider = (props) => {
  const [videos, setVideos] = useState([]);
  const [search, setSearch] = useState('wizeline');

  useEffect(() => {
    const getVideos = async () => {
      try {
        const url = `https://www.googleapis.com/youtube/v3/search?part=snippet&&key=${process.env.REACT_APP_GOOGLE_APP_API_KEY}&type=video&maxResults=23&q=${search}`;
        const response = await fetch(url);
        const data = await response.json();
        setVideos(data.items);
      } catch (error) {
        throw new Error(error);
      }
    };
    getVideos();
  }, [search]);

  return (
    <VideoListContext.Provider
      value={{
        setSearch,
        videos,
      }}
    >
      {props.children}
    </VideoListContext.Provider>
  );
};

export default VideoListProvider;
