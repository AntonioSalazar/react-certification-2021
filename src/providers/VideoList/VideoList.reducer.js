export const reducer = (state, action) => {
    switch (action.type) {
        case 'FETCH_SUCCESS':
            return {
                ...state,
                loading: false,
                posts: action.payload,
                error: ''
            }

        case 'FETCH_ERROR':
            return {
                ...state,
                loading: false,
                posts: [],
                error: 'Unable to fetch data!'
            }

        case 'ADD_TO_FAVORITES':
            const favoriteVideos = state.favoriteVideos.concat(action.payload)

            localStorage.setItem('favorite_videos', JSON.stringify(favoriteVideos))
            return {
                ...state,
                favoriteVideos
            }

        case 'REMOVE_FROM_FAVORITES':   
            const favoriteVideosFromLocalStorage = localStorage.getItem('favorite_videos');
            const arrayOfVideos = JSON.parse(favoriteVideosFromLocalStorage)
            const filteredVideos = arrayOfVideos.filter(video => video.id !== action.payload)
            localStorage.setItem('favorite_videos', JSON.stringify(filteredVideos))
            return {
                ...state,
                filteredVideos
            }
            
    
        default:
            return state
    }
}