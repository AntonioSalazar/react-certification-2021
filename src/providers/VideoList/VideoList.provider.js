import React, { createContext, useEffect, useState, useReducer } from 'react';
import { useHistory } from "react-router-dom";
import { reducer } from './VideoList.reducer';
import Swal from 'sweetalert2'
export const VideoListContext = createContext();

const VideoListProvider = (props) => {
  const [search, setSearch] = useState('wizeline');
  const initialState = {
    loading: true,
    posts: [],
    error: '',
    favoriteVideos: []
  };
  const [videosState, dispatch] = useReducer(reducer, initialState);
  const history = useHistory()
  useEffect(() => {
    let mounted  = true;
    const getVideos = async () => {
        try {
          const url = `https://www.googleapis.com/youtube/v3/search?part=snippet&&key=${process.env.REACT_APP_GOOGLE_APP_API_KEY}&type=video&maxResults=23&q=${search}`;
          const response = await fetch(url);
          if (mounted) {
            const data = await response.json();
            dispatch({ type: 'FETCH_SUCCESS', payload: data.items });
          }
          history.push('/')
        } catch (error) {
          dispatch({ type: 'FETCH_ERROR' });
          throw new Error(error);
        }
      
    };
    getVideos();
    return () => mounted = false;

  }, [search, history]);

  const addToFavorites = (video) => {
    dispatch({
      type: 'ADD_TO_FAVORITES',
      payload: video
    })
    Swal.fire(
      'Video added to your favorites'
    )
  };

  const removeFromFavorites = id => {
    dispatch({
      type: 'REMOVE_FROM_FAVORITES',
      payload: id
    })
    Swal.fire(
      'Video has been removed from favorites'
    )
  }

  return (
    <VideoListContext.Provider
      value={{
        setSearch,
        addToFavorites,
        videosState, 
        removeFromFavorites
      }}
    >
      {props.children}
    </VideoListContext.Provider>
  );
};

export default VideoListProvider;
