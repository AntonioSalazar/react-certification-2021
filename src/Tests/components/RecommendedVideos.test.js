import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import RecommendedVideos from '../../components/RecommendedVideos';
import { VideoListContext } from '../../providers/VideoList/VideoList.provider';



describe('Tests of the RecommendedVideos component', () => {
    const posts = [
        {
            "kind": "youtube#searchResult",
            "etag": "erqeM78PZDWIBe8qOGHGM2WdSE8",
            "id": 
                "nmXMgqjQzls"
            ,
            "snippet": {
                "publishedAt": "2019-09-30T23:54:32Z",
                "channelId": "UCPGzT4wecuWM0BH9mPiulXg",
                "title": "Video Tour | Welcome to Wizeline Guadalajara",
                "description": "Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...",
                "thumbnails": {
                    "default": {
                        "url": "https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg",
                        "width": 120,
                        "height": 90
                    },
                    "medium": {
                        "url": "https://i.ytimg.com/vi/nmXMgqjQzls/mqdefault.jpg",
                        "width": 320,
                        "height": 180
                    },
                    "high": {
                        "url": "https://i.ytimg.com/vi/nmXMgqjQzls/hqdefault.jpg",
                        "width": 480,
                        "height": 360
                    }
                },
                "channelTitle": "Wizeline",
                "liveBroadcastContent": "none",
                "publishTime": "2019-09-30T23:54:32Z"
            }
        },
        {
            "kind": "youtube#searchResult",
            "etag": "e69LZQ8KMVzNOm0RfQ1IjaUCbmw",
            "id": 
                 "HYyRZiwBWc8"
            ,
            "snippet": {
                "publishedAt": "2019-04-18T18:48:04Z",
                "channelId": "UCPGzT4wecuWM0BH9mPiulXg",
                "title": "Wizeline Guadalajara | Bringing Silicon Valley to Mexico",
                "description": "Wizeline continues to offer a Silicon Valley culture in burgeoning innovation hubs like Mexico and Vietnam. In 2018, our Guadalajara team moved into a ...",
                "thumbnails": {
                    "default": {
                        "url": "https://i.ytimg.com/vi/HYyRZiwBWc8/default.jpg",
                        "width": 120,
                        "height": 90
                    },
                    "medium": {
                        "url": "https://i.ytimg.com/vi/HYyRZiwBWc8/mqdefault.jpg",
                        "width": 320,
                        "height": 180
                    },
                    "high": {
                        "url": "https://i.ytimg.com/vi/HYyRZiwBWc8/hqdefault.jpg",
                        "width": 480,
                        "height": 360
                    }
                },
                "channelTitle": "Wizeline",
                "liveBroadcastContent": "none",
                "publishTime": "2019-04-18T18:48:04Z"
            }
        }
      ]

    const wrapper = mount(

        <VideoListContext.Provider
            value={{
                posts
            }}
            >
            <MemoryRouter>
                < RecommendedVideos videos={posts}/>
            </MemoryRouter>
        </VideoListContext.Provider>
    )

    

    test('should render the component correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    test('should display all of the videos passed', () => {
        let videos = wrapper.find('recommended-video__card')
        expect(videos.length).toBe(videos.length)
    })

})
