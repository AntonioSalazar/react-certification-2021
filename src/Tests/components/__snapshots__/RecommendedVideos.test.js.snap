// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of the RecommendedVideos component should render the component correctly 1`] = `
<div
  className="recommended-video__container"
>
  <a
    href="/favorite-video/[object Object]"
    onClick={[Function]}
    style={
      Object {
        "color": "#000",
        "textDecoration": "none",
      }
    }
  >
    <div
      className="recommended-video__card"
    >
      <div
        className="recommended-video__card-img"
      >
        <img
          alt="Video Tour | Welcome to Wizeline Guadalajara"
          src="https://i.ytimg.com/vi/nmXMgqjQzls/mqdefault.jpg"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
      <div
        className="recommended-video__card-text"
      >
        <span
          className="sc-bdnxRM eNjdHH"
        >
          Video Tour | Welcome to Wizeline Guadalajara
        </span>
      </div>
    </div>
    <hr
      style={
        Object {
          "color": "#060B26",
          "margin": "0 auto",
          "width": "80%",
        }
      }
    />
  </a>
  <a
    href="/favorite-video/[object Object]"
    onClick={[Function]}
    style={
      Object {
        "color": "#000",
        "textDecoration": "none",
      }
    }
  >
    <div
      className="recommended-video__card"
    >
      <div
        className="recommended-video__card-img"
      >
        <img
          alt="Wizeline Guadalajara | Bringing Silicon Valley to Mexico"
          src="https://i.ytimg.com/vi/HYyRZiwBWc8/mqdefault.jpg"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
      <div
        className="recommended-video__card-text"
      >
        <span
          className="sc-bdnxRM eNjdHH"
        >
          Wizeline Guadalajara | Bringing Silicon Valley to Mexico
        </span>
      </div>
    </div>
    <hr
      style={
        Object {
          "color": "#060B26",
          "margin": "0 auto",
          "width": "80%",
        }
      }
    />
  </a>
</div>
`;
